sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs

```diff
@@ -4101,6 +4115,152 @@ private void SetNameFieldsIfNull()
+        /// a failure occurs.
+        /// </remarks>
+        public Uri GenerateSasUri(
+            BlobSasBuilder builder)
+        {
+            builder = builder ?? throw Errors.ArgumentNull(nameof(builder));
+            builder.BlobContainerName = string.IsNullOrEmpty(builder.BlobContainerName) ? BlobContainerName : builder.BlobContainerName;
+            builder.BlobName = string.IsNullOrEmpty(builder.BlobName) ? Name : builder.BlobName;
+            if (!builder.BlobContainerName.Equals(BlobContainerName, StringComparison.InvariantCulture))
+            {
+                // TODO: throw proper exception for non-matching builder name
+                // e.g. containerName doesn't match or leave the containerName in builder
+                // should be left empty. Or should we always default to the client's ContainerName
+                // and chug along if they don't match?
+                throw Errors.SasNamesNotMatching(
+                    nameof(builder.BlobContainerName),
+                    nameof(BlobSasBuilder),
+                    nameof(BlobContainerName));
+            }
+            if (!builder.BlobName.Equals(Name, StringComparison.InvariantCulture))
+            {
+                // TODO: throw proper exception for non-matching builder name
+                // e.g. containerName doesn't match or leave the containerName in builder
+                // should be left empty. Or should we always default to the client's ContainerName
+                // and chug along if they don't match?
+                throw Errors.SasNamesNotMatching(
+                    nameof(builder.BlobName),
+                    nameof(BlobSasBuilder),
+                    nameof(Name));
+            }
+            UriBuilder sasUri = new UriBuilder(Uri);
+            sasUri.Query = builder.ToSasQueryParameters(_storageSharedKeyCredential).ToString();
+            return sasUri.Uri;
+        }
+
+        /// <summary>
+        /// The <see cref="GenerateUserDelegationSasUri"/> returns a Uri that
+        /// generates a User Delegation SAS based on the Client properties and builder passed.
+        ///
+        /// For more information, see
+        /// <see href="https://docs.microsoft.com/en-us/rest/api/storageservices/create-user-delegation-sas">
+        /// Constructing a User Delegation SAS</see>.
+        /// </summary>
+        /// <param name="builder">
+        /// Used to generate a Shared Access Signature (SAS).
+        /// </param>
+        /// <param name="delegationKey">
+        /// User Delegation Key used to generate the User Delegation SAS
+        /// </param>
+        /// <returns>
+        /// A <see cref="BlobSasBuilder"/> on successfully deleting.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public Uri GenerateUserDelegationSasUri(
```

Comment:
same here.

Reply:
I see your point on this. Because the customer still has to provide the `UserDelegationKey`. Yeah cause it would take the same amount of space as doing, `builder.ToSasQueryParameters(delegationKey, client.AccountName);`

Anyone else opposed to getting rid of this API? I know in Java they currently have it.

Reply:
Let's remove it for now.  A better experience might be `public Response<Uri> GenerateUserDelegationSasUriAsync(BlobSasBuilder builder)` that creates a `BlobServiceClient` with the same pipeline, calls `GetUserDelegationKey`, and uses that to generate a SAS token from the builder.  A lot of customers would have a problem with an API that doesn't make the request because there's no way to get a `BlobServiceClient` with the same pipeline publicly today (we only let you move down the hierarchy).

But all of that will take more discussion, so I'm fine having that conversation after we see how customers react to the other APIs here.